// biz9-data-app/index.js
/*
Copyright 2016 Certified CoderZ
Author: Brandon Poole Sr. (biz9framework@gmail.com)
License GNU General Public License v3.0
Description: BiZ9 Framework: Data-App
*/
// -- biz9 --
const {Str,Num,Obj,Log,Response_Field} = require("biz9-utility");
const {Config,Data_Config,Url,Table} = require("./constant");
const {Service} = require("./service");
const {Memory_Logic}= require('./memory');
// -- other --
const dayjs = require('dayjs');

class Data_Service {
    static ping = async (url)  => {
        const [biz_response,biz_data] = await Service.ping(url);
        return [biz_response,biz_data];
    }
    static count = async (url,search,option)  => {
        const [biz_response,biz_data] = await Service.search(url,search,option);
        return [biz_response,biz_data];
    }
    static copy = async (url,table,id,option)  => {
        const [biz_response,biz_data] = await Service.copy(url,table,id,option);
        return [biz_response,biz_data];
    }
    static delete = async (url,table,id,option)  => {
        const [biz_response,biz_data] = await Service.delete(url,table,id,option);
        return [biz_response,biz_data];
    }
    static delete_search = async (url,search,option)  => {
        const [biz_response,biz_data] = await Service.delete_search(url,search,option);
        return [biz_response,biz_data];
    }
    static get = async (url,table,id,option)  => {
        const [biz_response,biz_data] = await Service.get(url,table,id,option);
        return [biz_response,biz_data];
    }
    static post = async (url,table,id,post_data,option)  => {
        const [biz_response,biz_data] = await Service.post(url,table,id,post_data,option);
        return [biz_response,biz_data];
    }
    static post_items = async (url,items,option)  => {
        const [biz_response,biz_data] = await Service.post_items(url,items,option);
        return [biz_response,biz_data];
    }
    static search = async (url,search,option)  => {
        const [biz_response,biz_data] = await Service.search(url,search,option);
        return [biz_response,biz_data];
    }
}
class Data_Url {
    static PING="biz9/data/ping";
    static COPY="biz9/data/copy";
    static COUNT="biz9/data/count";
    static DELETE="biz9/data/delete";
    static DELETE_SEARCH="biz9/data/delete_search";
    static GET="biz9/data/get";
    static POST="biz9/data/post";
    static POST_ITEMS="biz9/data/post_items";
    static SEARCH="biz9/data/search";
}
class Data_Value_Type {
    static COUNT = 'count';
    static IMAGE = 'image';
    static ITEMS = 'items';
    static NOTE = 'note';
    static TEXT = 'text';
    static ONE = 'one';
}
class Data_Response_Field {
    static ITEM_DELETE_COUNT = 'item_delete_count';
    static ITEMS_COUNT_CONFIRM = 'items_count';
    static ITEM_COPY_CONFIRM = 'item_copy_confirm';
    static ITEM_COPY_FAIL = 'item_copy_fail';
    static ITEMS_SEARCH_CONFIRM = 'items_search_confirm';
    static ITEMS_SEARCH_FAIL = 'items_search_fail';
    static ITEMS_DELETE_COUNT = 'items_delete_count';
    static ITEMS_DELETE_CONFIRM = 'items_delete';
    static ITEMS_DELETE_FAIL = 'items_fail_delete';
    static ITEM_CACHE_DELETE = 'cache_delete';
    static ITEM_DATABASE_DELETE = 'database_delete';
    static ITEMS_POST_CONFIRM = 'items_post';
    static ITEMS_POST_FAIL = 'items_post_fail';
    static ITEM_GROUP_DELETE = 'group_delete';
    static OPTION_FIELD = 'option_field';
    static OPTION_GROUPS = 'option_groups';
    static PARAM_APP_ID = 'param_app_id';
    static PARAM_TABLE = 'param_table';
    static PARAM_ID = 'param_id';
    static PARAM_IDS = 'param_ids';
    static PARAM_PARENT_ID = 'param_parent_id';
    static PARAM_PARENT_TABLE = 'param_parent_table';
    static PARAM_USER_ID = 'param_user_id';
    static PARAM_DATA = 'param_data';
    static PARAM_URL = 'param_url';
    static PARAM_OPTION = 'param_option';

    static PARAM_OPTION_CACHE_DELETE = 'param_option_cache_delete';
    static PARAM_OPTION_TITLE = 'param_option_title';
    static PARAM_OPTION_JOINS = 'param_option_joins';
    static PARAM_OPTION_FIELD = 'param_option_field';
    static PARAM_OPTION_RESET = 'param_option_reset';
    static PARAM_OPTION_FOREIGNS = 'param_option_foreigns';
    static PARAM_OPTION_CLEAN = 'param_option_clean';

    static PARAM_SEARCH = 'param_search';
    static PARAM_FILTER = 'param_filter';
    static PARAM_SORT_BY = 'param_sort_by';
    static PARAM_PAGE_CURRENT = 'param_page_current';
    static PARAM_PAGE_SIZE = 'param_page_size';
    static PARAM_DATA_ITEMS = 'param_data_items';
    static PARAM_DATA_ITEMS_COUNT = 'param_data_items_count';
    static RESPONSE_ITEMS_COUNT = 'response_items_count';
    static RESPONSE_TOTAL_ITEMS_COUNT = 'response_total_items_count';
}
class Data_Field {
    static COPY='copy';
    static DATE_CREATE='date_create';
    static DATE_SAVE='date_save';
    static GROUP='group';
    static GROUP_ID='group_id';
    static ID='id';
    static _ID='_id';
    static PARENT = 'parent';
    static PARENT_ID = 'parent_id';
    static PARENT_TABLE='parent_table';
    static USER='user';
    static USER_ID = 'user_id';
    static USER_TABLE='user_table';
    static SOURCE='source';
    static SOURCE_ID='source_id';
    static SOURCE_TABLE='source_table';
    static TABLE = 'table';
    static TITLE='title';
    static TITLE_URL='title_url';
    //data_value_type
    static ITEM='item';
    static ITEMS='items';
    static ITEMS_FIELD = 'items_field';
    static ITEMS_TITLE = 'items_title';
    static ITEMS_VALUE = 'items_value';
    //cache
    static MINUTE = 'minute';
    static TIME = '_time';
    static DATA = '_data';
}
class Data_Type {
    //data_source
    static DATA_SOURCE_CLIENT="client";
    static DATA_SOURCE_CLIENT_CACHE="client_cache";
    static DATA_SOURCE_DATABASE="database";
    static DATA_SOURCE_NOT_FOUND="not_found";
    static DATA_SOURCE_SERVER="server";
    static DATA_SOURCE_SERVER_CACHE="server_cache";
    //other
    static JOIN="join";
    //result
    static RESULT_OK = 'resultOK';
    static RESULT_OK_DELETE = 'delete_resultOK';
    static RESULT_OK_DELETE_CACHE = 'delete_cache_resultOK';
    static RESULT_OK_DELETE_COUNT = 'delete_count';
    static RESULT_OK_DELETE_GROUP_COUNT = 'delete_group_count';
    static RESULT_OK_GROUP_DELETE = 'group_delete_resultOK';
    static RESULT_OK_DELETE_DATABASE = 'delete_database_resultOK';
    //source
    static SOURCE='source';
    static SOURCE_KEY='source_key';
    static SOURCE_TABLE='source_table';
    static SOURCE_ID='source_id';
    static SOURCE_PARENT_ID='source_parent_id';
    //sort_by
    static SORT_BY_ASC='asc';
    static SORT_BY_DESC='desc';
}
class Data_Title {
    //general
    static FOREIGN='Foreign';
    static GROUP='Group';
    static JOIN='Join';
    static N_A='N/A';
    //source
    static SOURCE_DATABASE='Database';
    static SOURCE_CACHE='Cache';
    static SOURCE_NOT_FOUND='Not-Found';
}
class Data_Table {
    static BLANK='blank_biz';
    static GROUP='group_biz';
    static SUB_VALUE='sub_value_biz';
    static USER='user_biz';
}
class Data_Logic {
    static get = (table,id,option) => {
        option = !Obj.check_is_empty(option) ? option : {};
        /* options
            - data
            - title
            - parent
            - user
            - generate_id
            - count
            */
        if(Str.check_is_null(table)){
            table = Data_Table.BLANK;
        }
        if(Str.check_is_null(id)){
            id = 0;
        }
        let data = {table:table,id:id};
        if(option.data){
            data = Obj.merge(data,option.data);
        }
        if(option.title){
            data[Data_Field.TITLE] = option.title;
        }
        if(option.parent){
            data[Data_Field.PARENT_TABLE] = option.parent[Data_Field.TABLE];
            data[Data_Field.PARENT_ID] = option.parent[Data_Field.ID];
        }
        if(option.user){
            data[Data_Field.USER_TABLE] = option.user[Data_Field.TABLE];
            data[Data_Field.USER_ID] = option.user[Data_Field.ID];
        }
        if(option.generate_id){
            data[Data_Field.ID] = Num.get_id(999);
        }
        if(option.count){
            let data_items = [];
            if(!data.title){
                data.title = 'Item';
            }
            for(let a = 1; a < option.count+1; a++){
                let copy = { ...data };
                copy.title = data.title + " " + a;
                if(option.generate_id){
                    copy[Data_Field.ID] = Num.get_id(999);
                }
                data_items.push(copy);
            }
            data = data_items;
        }
        return data;
    };
    static get_search = (table,filter,option) => {
        option = !Obj.check_is_empty(option) ? option : {sort_by:{},page_current:1,page_size:0};
        return {table:table,filter:filter,sort_by:option.sort_by?option.sort_by:{},page_current:option.page_current?option.page_current:1,page_size:option.page_size?option.page_size:0};
    };
    static get_join = (value_type,search,option) => {
        option = !Obj.check_is_empty(option)  ? option : {};
        value_type = value_type ? value_type : Data_Value_Type.ITEMS;
        search = search ? search : Data_Logic.get_search(Data_TABLE.BLANK);
        let field = option.field ? option.field : {};
        let distinct = option.distinct ? option.distinct : null; //distinct:{field:'title',sort_by:Data_Type.SEARCH_SORT_BY_DESC}
        let title = option.title ? Str.get_title_url(option.title) : Str.get_lower_case(Str.get_plural(Data_Title.JOIN));
        let foreigns = option.foreigns ? option.foreigns : [];

        search.page_current = !Str.check_is_null(option.page_current) ? option.page_current : search.page_current;
        search.page_size = !Str.check_is_null(option.page_size) ? option.page_size : search.page_size;
        return {value_type:value_type,search:search,field:field,title:title,distinct:distinct,foreigns:foreigns,page_current:search.page_current,page_size:search.page_size};
    }
    static get_foreign = (value_type,foreign_table,foreign_field,parent_field,option) => {
        option = !Obj.check_is_empty(option)  ? option : {};
        value_type = value_type ? value_type : Data_Value_Type.ITEMS;
        foreign_table = foreign_table ? foreign_table : {};
        foreign_field = foreign_field ? foreign_field : Data_Field.PARENT_ID;
        parent_field = parent_field ? parent_field : parent_field;
        let field = option.field ? option.field : null;
        let title = option.title ? Str.get_title_url(option.title) : Str.get_lower_case(Str.get_plural(Data_Title.FOREIGN));
        let page_current = option.page_current ? option.page_current : 1;
        let page_size = option.page_size ? option.page_size : 0;
        let sort_by = option.sort_by ? option.sort_by : {};
        let foreigns = option.foreigns ? option.foreigns : [];
        let not_found = option.not_found ? option.not_found : Data_Logic.get_not_found(foreign_table,0) ;
        return {not_found:not_found,value_type:value_type,sort_by:sort_by,foreign_table:foreign_table,foreign_field:foreign_field,parent_field:parent_field,field:field,title:title,page_current:page_current,page_size:page_size,foreigns:foreigns};
    };
    static get_foreign_parent = (parent_table,option) => {
        option = !Obj.check_is_empty(option)  ? option : {};
        option = Obj.merge(option,{title:'parent'});
        return Data_Logic.get_foreign(Data_Value_Type.ONE,parent_table,Data_Field.ID,Data_Field.PARENT_ID,option);
    };
    static get_message_by_response_field = (response_field) =>{
        switch(response_field){
            case Data_Response_Field.ITEMS_SEARCH_CONFIRM:
                return "Item(s) search success.";
                break;
            case Data_Response_Field.ITEMS_SEARCH_FAIL:
                return "Item(s) search failed.";
                break;
            case Response_Field.GET_CONFIRM:
                return "Item get successfully.";
                break;
            case Response_Field.GET_FAIL:
                return "Item get failed.";
                break;
            case Data_Response_Field.ITEM_COPY_CONFIRM:
                return "Item copied successfully.";
                break;
            case Data_Response_Field.ITEM_COPY_FAIL:
                return "Item copy failed.";
                break;
            case Data_Response_Field.ITEMS_COUNT_CONFIRM:
                return "Item(s)) counted successfully.";
                break;
            case Response_Field.POST_CONFIRM:
                return "Item posted successfully.";
                break;
            case Response_Field.POST_FAIL:
                return "Item posted failed.";
                break;
            case Data_Response_Field.ITEMS_POST_CONFIRM:
                return "Item(s) posted successfully.";
                break;
            case Data_Response_Field.ITEMS_POST_FAIL:
                return "Item(s) posted failed.";
                break;
            case Response_Field.DELETE_CONFIRM:
                return "Item deleted successfully.";
                break;
            case Response_Field.DELETE_FAIL:
                return "Item not deleted.";
                break;
            case Data_Response_Field.ITEMS_DELETE_FAIL:
                return "Item(s) not deleted.";
                break;
            case Data_Response_Field.ITEMS_DELETE_CONFIRM:
                return "Item(s) deleted successfully.";
                break;
            default:
                if(response_field){
                    return response_field;
                }else{
                    return "Error Message Not Availble";
                }
        }
    }
    static copy = (table,item,option)=>{
        let copy_item = Data_Logic.get(table,0);
        option = !Obj.check_is_empty(option)  ? option : {};
        const keys = Object.keys(item);
        keys.forEach(key => {
            if(
                key!=Data_Field.ID&&
                key!=Data_Field.TABLE&&
                key!=Data_Field.SOURCE&&
                key!=Data_Field.DATE_CREATE&&
                key!=Data_Field.DATE_SAVE&&
                key!=Data_Field.OBJ&&
                key!=Data_Field.USER&&
                key!=Data_Field.GROUP&&
                key!=Data_Field.ITEM&&
                !Obj.check_is_array(item[key])&&
                Obj.check_is_value(item[key])
            )
            {
                copy_item[key]=item[key];
            }
        });
        return copy_item;
    };
    static get_not_found = (table,id,option) =>{
        option=!Obj.check_is_empty(option)?option:{};
        if(!table){
            table = Data_Table.BLANK;
        }
        if(!id){
            id = 0;
        }
        let item = Data_Logic.get(table,id);
        item.id = 0;
        item.id_key = id;
        item.title = "Not Found";
        item.source = Data_Title.SOURCE_NOT_FOUND;
        return item;
    };
    static get_table_by_name = (table,option) =>{
        table = table?table:"";
        let table_title = "";
        option = option ? option : {lowercase:false,plural:false,url:false};
        table_title = String(Str.get_title(table.replaceAll('_',' ').replace('biz',''))).trim();
        if(option.plural){
            table_title = table_title + 's';
        }
        if(option.lowercase){
            table_title = table_title.toLowerCase();
        }
        if(option.url){
            table_title = table_title.replace(" ","_").toLowerCase();
        }
        return table_title;
    };
    static get_test_cost = () =>{
        return String(Str.get_id(999)) + "." + String(Str.get_id(99));
    }
    static get_test_parents = (count,option) =>{
        if(!count){
            count = 9;
        }
        let parents = [];
        for(let a = 0; a < count; a++){
            parents.push(Data_Logic.get_test_parent(option));
        }
        return parents;
    };
    static get_test_parent = (option) =>{
        option = !Obj.check_is_empty(option) ? option : {title:'Product '+Str.get_id(999)};
        let data = Data_Logic.get(Table.PRODUCT,0,option);
        data.cost = Data_Logic.get_test_cost();
        data.old_cost = Data_Logic.get_test_cost();
        data.category = "Category "+String(Str.get_id());
        data.type = "Type "+String(Str.get_id());
        data.sub_type = "Sub Type "+String(Str.get_id());
        data.stock = String(Str.get_id(3-1));
        data.tag = "Tag "+ Str.get_id() + ", Tag "+Str.get_id() + ", Tag "+ Str.get_id();
        return data;
    };
}
class Storage_Logic {
    static post = (key,value,option) => {
        option = !Obj.check_is_empty(option) ? option : {};
        if(!key){
            key = '';
        }
        key = Str.get_title_url(key);
        if(option.time_limit){
            Memory_Logic.post(key+Data_Field.TIME,dayjs().add(option.time_limit,Data_Field.MINUTE));
            Memory_Logic.post(key+Data_Field.DATA,value);
        }else{
            Memory_Logic.post(key,value);
        }
    }
    static delete_all = () => {
        Memory_Logic.delete_all();
    }
    static delete = (key) => {
        if(!key){
            key = '';
        }
        key = Str.get_title_url(key);
        Memory_Logic.delete(key);
        Memory_Logic.delete(key+Data_Field.TIME);
        Memory_Logic.delete(key+Data_Field.TIME);
    }
    static get = (key,option) => {
        option = !Obj.check_is_empty(option) ? option : {};
        if(!key){
            key = '';
        }
        key = Str.get_title_url(key);
        if(option.time_limit){
            let time = Memory_Logic.get(key+Data_Field.TIME);
            if(dayjs(time).isValid() && dayjs().isBefore(dayjs(time).add(option.time_limit))){
                return Memory_Logic.get(key+Data_Field.DATA);
            }else{
                return null;
            }
        }else{
            return Memory_Logic.get(key);
        }
    }
}

module.exports = {
    Data_Field,
    Data_Logic,
    Data_Response_Field,
    Data_Table,
    Data_Title,
    Data_Type,
    Data_Url,
    Data_Value_Type,
    Storage_Logic,
    Data_Service
};
